python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py

```diff
@@ -0,0 +1,204 @@
+                    " type of previous batches!"
+                )
+
+        while self.size >= self.saturation_level:
+            self.flush()
+
+    def flush(self, skip_partition_size_check=False) -> None:
+        """
+        Computes all uncomputed batches
+        """
+        min_batch_id = self.__batches[self.batch_col_name].min()
+        if isinstance(self.__batches, dask_cudf.DataFrame):
+            min_batch_id = min_batch_id.compute()
+        min_batch_id = int(min_batch_id)
+
+        partition_size = self.batches_per_partition * self.batch_size
+        partitions_per_saturation_level = self.saturation_level // partition_size
+
+        if skip_partition_size_check:
+            npartitions = int(ceil(len(self.__batches) / partition_size))
+        else:
+            npartitions = partitions_per_saturation_level
+
+        max_batch_id = min_batch_id + npartitions * self.batches_per_partition - 1
+        batch_id_filter = self.__batches[self.batch_col_name] <= max_batch_id
+
+        sample_fn = (
+            cugraph.uniform_neighbor_sample
+            if isinstance(self.__graph._plc_graph, pylibcugraph.graphs.SGGraph)
+            else cugraph.dask.uniform_neighbor_sample
+        )
+
+        # TODO semaphore check to prevent concurrent calls to uniform_neighbor_sample
+        samples = sample_fn(
+            self.__graph,
+            **self.__sample_call_args,
+            start_list=self.__batches[self.start_col_name][batch_id_filter],
+            batch_id_list=self.__batches[self.batch_col_name][batch_id_filter],
+            with_edge_properties=True,
+        )
+
+        self.__batches = self.__batches[~batch_id_filter]
+        if len(self.__batches) == 0:
+            self.__batches = None
+
+        self.__write(samples, min_batch_id, npartitions)
+
+    def __write(
+        self,
+        samples: Union[cudf.DataFrame, dask_cudf.DataFrame],
+        min_batch_id: int,
+        npartitions: int,
+    ) -> None:
+        # Ensure each rank writes to its own partition so there is no conflict
+        outer_partition = f"rank={self.__rank}"
+        outer_partition_path = os.path.join(self.__output_path, outer_partition)
+        os.makedirs(outer_partition_path, exist_ok=True)
+
+        for partition_k in range(npartitions):
+            ix_partition_start_inclusive = (
+                min_batch_id + partition_k * self.batches_per_partition
+            )
+            ix_partition_end_inclusive = (
+                min_batch_id + (partition_k + 1) * self.batches_per_partition - 1
+            )
+
+            inner_path = os.path.join(
+                outer_partition_path,
+                f"batch={ix_partition_start_inclusive}-{ix_partition_end_inclusive}",
+            )
+
+            f = (samples.batch_id >= ix_partition_start_inclusive) & (
+                samples.batch_id <= ix_partition_end_inclusive
+            )
+            samples[f].to_parquet(
+                inner_path,
```

Comment:
So we also seem to be writing empty partitions here. I suggest having a check like 
```suggestion
           if len(samples[f])==0:
              break

            samples[f].to_parquet(
                inner_path,
```